### DISCUSSION TITLE
Trigerring Test Method from normal function. Processing each CSV record for test

### DISCUSSION THREAD

**Discussion by narendarbhogala** [seleniumbase/SeleniumBase#2011] [Q&A] [answered]
Hi, 
I have a requirement, where I'm reading each row from the csv file and will have to perform test_method execution on each record to generate one test record/status for each record.
I created an csv read generator function in Common library file and called the processor function in my test_file. From the processor method after retrieving each row records, trying to call the test_method within the processor_method.
Now when I execute, it displays me error, i cannot call Test_methods and by default, it executes the test method first.
I tried to call the processor_method from main function but still it executes the test_method first.

Is there a way, i can create a test result for each record after reading it from the csv file.

Alternately, i made the processor_method as test_processor_method and called the required test_method as normal method_name and it runs for all the records in the csv file and creates one test result for me. But for any assertion failure for any record, it either fails the entire test or when i handle the assertion it just moves to next record with warning on console. but I wanted to create a test failure for each record if assertion fails and move to next record as new test call. 

Is there a way i can accomplish this?

**Comment by mdmintz** [comment, marked as answer]
The output of `pytest --co -q` is a test address, such as `basic_test.py::MyTestClass::test_basics`, which can be called directly with `pytest`. If you put each row into a CSV file, you can easily map tests to results. So if you build a CSV parser, you can pull the test address, run it with `pytest`, and then put the result back into the CSV file. This seems kind of off-topic for SeleniumBase, (since it is directly tied to `pytest` at that point), but this should point you in the right direction.

**Comment by narendarbhogala** [comment]
Can I do it with SeleniumBase? instead of pytest. As I'm trying to build the test framework with SeleniumBase, wanted to know how to get the results, like Pass or Fail for each test iteration of the test method for a single record. I'm new in building this and wanted to understand if this can be possible

**Reply by mdmintz** [reply]
SeleniumBase is a pytest plugin, so it's really pytest that runs and manages the tests. SeleniumBase takes care of the browser automation activities while pytest runs tests and structures results.
Eg. `pytest --html=report.html` for an html report after tests complete.
And the dashboard: `pytest --dashboard`.
Or both the report and the dashboard: `pytest --html=report.html --dashboard`.

**Reply by narendarbhogala** [reply]
May be I didn't phrase the question properly. I really appreciate the response and wanted insight on how to resolve.
**Initially I wrote two methods in my common library:**

  **def csv_read(self, csv_file_path, col_mapping):**
      **"""generator to read the values from csv file"""**

      with open(csv_file_path, 'r') as csv_file: #Opens CSV file in read mode
          csv_reader = csv.reader(csv_file)
          next(csv_reader, None) #Skips the header row in the CSV file

          for row in csv_reader: # for each row in the CSV file, fetches values in the entire row
              if len(row) >= max(col_mapping) +1 :
                  values = [row[col_index] for col_index in col_mapping]
                  yield values
 

  **def process_csv_rows(self, csv_file_path, col_mapping, processor_function):**
      **"""Calls the generator function to loop through each row from data sheet """**

      self.csv_generator = self.csv_read(csv_file_path, col_mapping)
      for values in self.csv_generator:
          self.refresh()
          processor_function(values)
          
          
**And in my test_file**

**def test_ViewDetails_DV_rows(self):**
  **"""Calls the generator function to loop through each row from the bulk upload data sheet and processes actions on each row"""**
  
      self.process_csv_rows(self.csv_file_path, self.col_mapping, **self.processor_function**)

**def processor_function():**
**"""This contains steps/actions to perform on the values fetched from the csv file"""**
assert_equal()
assertions()

**Doing this way, if the assertions fail, if one assertion fails entire test is failed. If I handle assertions in try, except, only one report is generated as my test_function is called only once.

So I I wanted to modify the code in such a way:**

**@pytest.fixture()
def ViewDetails_DV_rows(self):**
  **"""Calls the generator function to loop through each row from the bulk upload data sheet and processes actions on each row"""**
  
      self.process_csv_rows(self.csv_file_path, self.col_mapping)

**def test_processor_function(ViewDetails_DV_rows):**
**"""This contains steps/actions to perform on the values fetched from the csv file"""**
          assert_equal()
          assertions()

**I tried pytest.mark.parameterize also , but some how not able to efficiently pass the required fetched csv row to the test method. When using pytest fixures.

TypeError: test_processor_function() missing 1 required positional argument: 'ViewDetails_DV_rows'

 I'm not sure how to process the test_method for each record fetched from the csv read file.**

**Reply by mdmintz** [reply]
If using pytest fixtures with SeleniumBase, you need to use one of the formats that has the `sb` fixture. See https://github.com/seleniumbase/SeleniumBase/blob/master/help_docs/syntax_formats.md (Eg. Format 3 or Format 4)

**Reply by narendarbhogala** [reply]
without using fixures, how to invoke test_method from csv_read method?
As it is not invoking the test_method from within normal method (csv_read)

def csv_read():
            retrieve each row from csv read
            call test_method() for each row

def test_method():
             assertions on each row
